python/ql/src/semmle/python/frameworks/Invoke.qll

```diff
@@ -16,102 +17,44 @@ private module Invoke {
-  }
-
-  /**
-   * Gets a reference to the attribute `attr_name` of the `invoke` module.
-   * WARNING: Only holds for a few predefined attributes.
-   */
-  private DataFlow::Node invoke_attr(string attr_name) {
-    result = invoke_attr(DataFlow::TypeTracker::end(), attr_name)
-  }
+  API::Node invoke() { result = API::moduleImport("invoke") }
 
   /** Provides models for the `invoke` module. */
   module invoke {
     /** Gets a reference to the `invoke.context` module. */
-    DataFlow::Node context() { result = invoke_attr("context") }
+    API::Node context() { result = invoke().getMember("context") }
 
     /** Provides models for the `invoke.context` module */
     module context {
       /** Provides models for the `invoke.context.Context` class */
       module Context {
         /** Gets a reference to the `invoke.context.Context` class. */
-        private DataFlow::Node classRef(DataFlow::TypeTracker t) {
-          t.start() and
-          result = DataFlow::importNode("invoke.context.Context")
-          or
-          t.startInAttr("Context") and
-          result = invoke::context()
-          or
-          // handle invoke.Context alias
-          t.start() and
-          result = invoke_attr("Context")
-          or
-          exists(DataFlow::TypeTracker t2 | result = classRef(t2).track(t2, t))
+        API::Node classRef() {
+          result =
+            [API::moduleImport("invoke").getMember("context"), API::moduleImport("invoke")]
+                .getMember("Context")
         }
 
-        /** Gets a reference to the `invoke.context.Context` class. */
-        DataFlow::Node classRef() { result = classRef(DataFlow::TypeTracker::end()) }
-
         /** Gets a reference to an instance of `invoke.context.Context`. */
-        private DataFlow::Node instance(DataFlow::TypeTracker t) {
+        private DataFlow::LocalSourceNode instance(DataFlow::TypeTracker t) {
           t.start() and
-          result.asCfgNode().(CallNode).getFunction() =
-            invoke::context::Context::classRef().asCfgNode()
-          or
-          t.start() and
-          exists(Function func |
-            func.getADecorator() = invoke_attr("task").asExpr() and
-            result.(DataFlow::ParameterNode).getParameter() = func.getArg(0)
+          (
+            result = invoke::context::Context::classRef().getACall()
+            or
+            exists(Function func |
+              func.getADecorator() = invoke().getMember("task").getAUse().asExpr() and
+              result.(DataFlow::ParameterNode).getParameter() = func.getArg(0)
+            )
           )
           or
           exists(DataFlow::TypeTracker t2 | result = instance(t2).track(t2, t))
         }
 
         /** Gets a reference to an instance of `invoke.context.Context`. */
-        DataFlow::Node instance() { result = instance(DataFlow::TypeTracker::end()) }
+        DataFlow::LocalSourceNode instance() { result = instance(DataFlow::TypeTracker::end()) }
```

Comment:
Hmm, I'm not too sure I like this change, since it changes what the predicate does, compared to other `instance()` predicates in other modules/qll files. I can see that in this context, it wouldn't actually change the _end result_.

Would it be ok to dumb this down to using `DataFlow::Node`, and then do _one_ swooping change over all our library modeling in the future changing all `instance()` predicates to have `DataFlow::LocalSourceNode` results?

```suggestion
        DataFlow::Node instance() { instance(DataFlow::TypeTracker::end()).flowsTo(result) }
```

Reply:
You're right. For our purposes, `LocalSourceNode` is sufficient (since we only ever use it in another type tracker), but external users may be caught unawares. I will add a change note explaining how to get the old behaviour back (by using `flowsTo`).

Reply:
Wait a minute... Is this predicate even externally visible? It's inside `private module Invoke`, after all...

Reply:
yes, it's not externally visible. And I'm not really too much against changing my idea about what `instance()` is going to do, I just think it would nice to do over all our library modeling in _one_ PR, so we don't end up in a situation where _some_ `instance()` predicates behave differently than others.

Reply:
This is an excellent point. :+1: I could easily see myself getting confused by this.